Content.Server/VendingMachines/VendingMachineSystem.cs

```diff
@@ -0,0 +1,243 @@
+using System.Collections.Generic;
+using Content.Shared.Interaction;
+using Robust.Shared.Audio;
+using Robust.Shared.GameObjects;
+using Robust.Shared.IoC;
+using Robust.Shared.Player;
+using System;
+using System.Linq;
+using Content.Server.Popups;
+using Content.Server.Power.Components;
+using Content.Server.WireHacking;
+using Content.Shared.Access.Components;
+using Content.Shared.Access.Systems;
+using Content.Shared.VendingMachines;
+using Robust.Server.GameObjects;
+using Robust.Shared.Localization;
+using Robust.Shared.Prototypes;
+using Robust.Shared.Random;
+using Content.Server.Advertise;
+using Content.Server.Throwing;
+using Robust.Shared.Maths;
+using static Content.Shared.VendingMachines.SharedVendingMachineComponent;
+
+namespace Content.Server.VendingMachines.systems
+{
+    public sealed class VendingMachineSystem : EntitySystem
+    {
+        [Dependency] private readonly AdvertiseSystem _advertisementSystem = default!;
+        [Dependency] private readonly IRobustRandom _random = default!;
+        [Dependency] private readonly IPrototypeManager _prototypeManager = default!;
+
+        public override void Initialize()
+        {
+            base.Initialize();
+            SubscribeLocalEvent<VendingMachineComponent, ComponentInit>(OnComponentInit);
+            SubscribeLocalEvent<VendingMachineComponent, ActivateInWorldEvent>(HandleActivate);
+            SubscribeLocalEvent<VendingMachineComponent, PowerChangedEvent>(OnPowerChanged);
+        }
+
+        private void OnComponentInit(EntityUid uid, VendingMachineComponent component, ComponentInit args)
+        {
+            base.Initialize();
+
+            if (component.UserInterface != null)
+            {
+                component.UserInterface.OnReceiveMessage += component.OnUiReceiveMessage;
+            }
+            if (EntityManager.TryGetComponent(component.Owner, out ApcPowerReceiverComponent? receiver))
+            {
+                TryUpdateVisualState(component, receiver.Powered ? VendingMachineVisualState.Normal : VendingMachineVisualState.Off);
+            }
+
+            InitializeFromPrototype(component);
+        }
+
+        private void HandleActivate(EntityUid uid, VendingMachineComponent component, ActivateInWorldEvent args)
+        {
+            if (!component.Powered ||
+                !EntityManager.TryGetComponent(args.User, out ActorComponent? actor))
+            {
+                return;
+            }
+            var wires = EntityManager.GetComponent<WiresComponent>(component.Owner);
```

Comment:
Don't use  ` EntityManager.GetComponent` if you are not 100% sure that component exist. Use `TryComp()`.

Reply:
To expand on this, the only time that you're 100% sure that a component exists is with components that have been checked using HasComponent or the component is Transform or Metadata. But if you're fetching the component for any of those, then it's better to use `TryComp()`, `Transform()` and `Metadata()` respectively.